ouroboros-network/src/Ouroboros/Network/BlockFetch/Decision.hs

```diff
@@ -547,6 +564,48 @@ filterNotAlreadyInFlightWithOtherPeers FetchModeBulkSync chains =
             _other                  -> peerFetchBlocksInFlight inflight
         | (_, status, inflight, _) <- chains ]
 
+    -- The highest slot number that is or has been in flight for any peer.
+    maxSlotNoInFlightWithOtherPeers = maximum $ Nothing :
+      [ peerFetchMaxSlotNo inflight | (_, _, inflight, _) <- chains ]
+
+-- | Filter a fragment, but stop the filtering (and add the remaining fragment
+-- untouched to the list of fragments, just like
+-- 'ChainFragment.filterWithStop') when a header's slot number is greater than
+-- the given slot number (@maxSlotNo@).
+--
+-- This slot number should correspond to some block or header after which it
+-- no longer makes sense to keep filtering, as there will be no more elements
+-- to remove.
+--
+-- PRECONDITION: for all @hdr@ in the chain fragment: if @blockSlot hdr >
+-- maxSlotNo@ then the predicate should not hold for any header after @hdr@ in
+-- the chain fragment.
+--
+-- For example, when filtering out already downloaded blocks from the
+-- fragment, it does not make sense to keep filtering after having encountered
+-- the highest slot number the ChainDB has seen so far, all blocks with a
+-- greater slot number cannot not have been downloaded yet. When the candidate
+-- fragments get far ahead of the current chain, e.g., @2k@ headers, this
+-- optimisation avoids the linear cost of filtering these headers when we know
+-- in advance they will all remain in the final fragment.
+--
+-- In case the given slot number is 'Nothing', no filtering takes place, as
+-- there should be no matches.
+--
+-- For example, when filtering out blocks already in-flight for the given
+-- peer, the given @maxSlotNo@ can correspond to the block with the highest
+-- slot number that so far has been in-flight for the given peer. When no
+-- blocks have been in-flight yet, @maxSlotNo@ can be 'Nothing', in which case
+-- no filtering needs to take place, which makes sense, as there are no blocks
+-- to filter out.
```

Comment:
Let's add a sentence here explaining that this is conservative, and that the case where blocks (for whatever reason) need to be in flight more than once would okay.